go.mod

```diff
@@ -0,0 +1,98 @@
+module github.com/openshift/hive
+
+go 1.13
+
+require (
+	github.com/Azure/azure-sdk-for-go v38.1.0+incompatible
+	github.com/Azure/go-autorest/autorest v0.9.3
+	github.com/Azure/go-autorest/autorest/azure/auth v0.4.1
+	github.com/Azure/go-autorest/autorest/to v0.3.1-0.20191028180845-3492b2aff503
+	github.com/aws/aws-sdk-go v1.29.24
+	github.com/docker/go-healthcheck v0.1.0
+	github.com/docker/go-units v0.4.0
+	github.com/evanphx/json-patch v4.5.0+incompatible
+	github.com/ghodss/yaml v1.0.1-0.20190212211648-25d852aebe32
+	github.com/golang/mock v1.3.1
+	github.com/google/gofuzz v1.1.0
+	github.com/google/uuid v1.1.1
+	github.com/heptio/velero v1.0.0
+	github.com/json-iterator/go v1.1.8
+	github.com/miekg/dns v1.1.15
+	github.com/modern-go/reflect2 v1.0.1
+	github.com/onsi/gomega v1.8.1
+	github.com/openshift/api v3.9.1-0.20191111211345-a27ff30ebf09+incompatible
+	github.com/openshift/cluster-api v0.0.0-20191129101638-b09907ac6668
+	github.com/openshift/cluster-api-provider-gcp v0.0.1-0.20200120152131-1b09fd9e7156
+	github.com/openshift/cluster-autoscaler-operator v0.0.0-20190521201101-62768a6ba480
+	github.com/openshift/cluster-network-operator v0.0.0-20190207145423-c226dcab667e
+	github.com/openshift/generic-admission-server v1.14.0
+	github.com/openshift/installer v0.9.0-master.0.20200331173433-d200b241c184
+	github.com/openshift/library-go v0.0.0-20200210105614-4bf528465627
+	github.com/pkg/errors v0.9.1
+	github.com/prometheus/client_golang v1.2.1
+	github.com/sirupsen/logrus v1.4.2
+	github.com/spf13/cobra v0.0.5
+	github.com/spf13/pflag v1.0.5
+	github.com/stretchr/testify v1.4.0
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b // indirect
+	golang.org/x/net v0.0.0-20200226121028-0de0cce0169b
+	golang.org/x/oauth2 v0.0.0-20191202225959-858c2ad4c8b6
+	golang.org/x/tools v0.0.0-20200331202046-9d5940d49312 // indirect
+	google.golang.org/api v0.14.0
+	gopkg.in/ini.v1 v1.51.0
+	gopkg.in/yaml.v2 v2.2.8
+	k8s.io/api v0.17.2
+	k8s.io/apiextensions-apiserver v0.17.1
+	k8s.io/apimachinery v0.17.3
```

Comment:
Can you clarify how you accomplished this? I'm confused by the mix of 1.17 and 1.14 versions for Kube things here, but then we seem to override to 1.14 with replace lines below?

Was there any prior art you followed to pull this off?

Reply:
That's a good question. The `require` stanza was entirely generated by the go tooling. That is the list of modules required by the code and its dependencies. When I tried making manual changes to that, the go tooling would overrule me with what it discovered. It was generated through a couple go commands.
```
go mod init github.com/openshfit/hive
go get -d github.com/openshfit/installer@d200b241c1844edc8609d1865bf9eeda1de95f85
go get -d github.com/openshift/cluster-network-operator@c226dcab667ea57774c81b73102f08bff5c2d518
go mod tidy
make build
make test
go mod tidy
```

But, I do see that I can make some changes. Let me play with it to see if I can get the requires to show the versions that we want.

Reply:
Ultimately, even after setting everything to the versions that we want in the `require` stanza, the go tooling fills it back in with other versions that it finds as upgraded versions needed by dependencies.

Reply:
So it's essentially require automatically determining that we need things like kube 1.17, but then we override with repalcements and force it back to 1.14? I'm slightly shocked this is compiling but that's great that it is.

Reply:
Right.

For example,
```
$ go list -m -json k8s.io/api
{
	"Path": "k8s.io/api",
	"Version": "v0.17.2",
	"Replace": {
		"Path": "k8s.io/api",
		"Version": "v0.0.0-20190409021203-6e4e0e4f393b",
		"Time": "2019-04-09T02:12:03Z",
		"Dir": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b",
		"GoMod": "/home/staebler/go/pkg/mod/cache/download/k8s.io/api/@v/v0.0.0-20190409021203-6e4e0e4f393b.mod"
	},
	"Dir": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b",
	"GoMod": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b/go.mod"
}
```

Reply:
And it goes through all of the following to determine the latest version for the require.
```
$ go mod graph | grep "k8s.io/api@"
github.com/openshift/hive k8s.io/api@v0.17.2
github.com/openshift/machine-api-operator@v0.2.1-0.20191128180243-986b771e661d k8s.io/api@v0.0.0-20190918195907-bd6ac527cfd2
github.com/openshift/machine-api-operator@v0.2.1-0.20200310180732-c63fa2b143f0 k8s.io/api@v0.17.2
github.com/openshift/installer@v0.9.0-master.0.20200331173433-d200b241c184 k8s.io/api@v0.17.2
github.com/openshift/client-go@v0.0.0-20191216194936-57f413491e9e k8s.io/api@v0.17.0
github.com/openshift/cluster-api-provider-ovirt@v0.1.1-0.20200128081049-840376ca5c09 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/client-go@v0.0.0-20200116152001-92a2713fa240 k8s.io/api@v0.17.1
github.com/openshift/cluster-api@v0.0.0-20191129101638-b09907ac6668 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/machine-config-operator@v4.2.0-alpha.0.0.20190917115525-033375cbe820+incompatible k8s.io/api@v0.17.1
github.com/openshift/cluster-api-provider-libvirt@v0.2.1-0.20191219173431-2336783d4603 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
github.com/openshift/runtime-utils@v0.0.0-20191011150825-9169de69ebf6 k8s.io/api@v0.0.0-20191003035645-10e821c09743
github.com/openshift/cluster-etcd-operator@v0.0.0-alpha.0.0.20191025163650-5854b5c48ce4 k8s.io/api@v0.0.0-20190313235455-40a48860b5ab
github.com/openshift/library-go@v0.0.0-20190619114638-6b58b672ee58 k8s.io/api@v0.17.0
github.com/cri-o/cri-o@v1.9.0-beta.2.0.20191003162030-4775e1c05c26 k8s.io/api@v0.0.0
github.com/containers/buildah@v1.11.2 k8s.io/api@v0.0.0-20190813020757-36bff7324fb7
github.com/containers/libpod@v1.5.1 k8s.io/api@v0.0.0-20190624085159-95846d7ef82a
github.com/openshift/library-go@v0.0.0-20200210105614-4bf528465627 k8s.io/api@v0.17.0
github.com/openshift/cluster-api@v0.0.0-20191030113141-9a3a7bbe9258 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/library-go@v0.0.0-20191003152030-97c62d8a2901 k8s.io/api@v0.17.0
github.com/hashicorp/vault-plugin-auth-kubernetes@v0.5.2-0.20190925162726-2e5b0b8184e6 k8s.io/api@v0.0.0-20190409092523-d687e77c8ae9
github.com/openshift/cluster-api-provider-gcp@v0.0.1-0.20200120152131-1b09fd9e7156 k8s.io/api@v0.0.0-20191003000013-35e20aa79eb8
k8s.io/kubelet@v0.17.1 k8s.io/api@v0.17.1
k8s.io/kubectl@v0.0.0-20200124035537-9f7d91504e51 k8s.io/api@v0.0.0-20200124032216-924612ff3bca
sigs.k8s.io/controller-runtime@v0.0.0-20190520212815-96b67f231945 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-tools@v0.2.2-0.20190919191502-76a25b63325a k8s.io/api@v0.0.0-20190222213804-5cb15d344471
sigs.k8s.io/controller-runtime@v0.2.0 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-runtime@v0.2.0-beta.2 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/cluster-api-provider-aws@v0.0.0 k8s.io/api@v0.17.2
sigs.k8s.io/cluster-api-provider-azure@v0.0.0 k8s.io/api@v0.0.0-20191121015604-11707872ac1c
sigs.k8s.io/controller-tools@v0.2.4 k8s.io/api@v0.0.0-20190222213804-5cb15d344471
sigs.k8s.io/controller-runtime@v0.4.0 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-runtime@v0.3.1-0.20191016212439-2df793d02076 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-tools@v0.2.2-0.20190930215132-4752ed2de7d2 k8s.io/api@v0.0.0-20190222213804-5cb15d344471
```